Add docs for upgrading from latest/edge to 2/edge. Add config option to allow changing the OIDC subject claim (#748)
canonical/synapse-operator: docs/how-to/migrate-to-mas.md

# How to back up and restore Synapse

> This document shows how to migrate a legacy synapse homeserver (`latest/edge` track) to an oidc-native homeserver (`2/edge` track).

Reviewer: ```suggestion
This document shows how to migrate a legacy Synapse homeserver (`latest/edge` track) to an oidc-native homeserver (`2/edge` track).
```

> This document shows how to migrate a legacy synapse homeserver (`latest/edge` track) to an oidc-native homeserver (`2/edge` track).

Reviewer 2: latest/edge -> 1/edge

Author: thanks, updated!

> This document shows how to migrate a legacy synapse homeserver (`latest/edge` track) to an oidc-native homeserver (`2/edge` track).

Reviewer 2: What about adding a warning to create a backup and mentioning that this procedure is irreversible? Also, the current instance should be stopped during the process.

Author: Hmm, I'm not sure that this process is irreversible as we don't touch the existing database, we can always remove the new charm and start over again.

## Initial deployment setup 

> This document will cover the migration path for a synapse charm on the `latest\edge` track, using the `saml-integrator` charm to provide authentication users via SAML.

Reviewer: ```suggestion
This document will cover the migration path for a synapse charm on the `latest\edge` track, using the `saml-integrator` charm to provide authentication via SAML.
```

## Prepare the new synapse charm
### Deploy the 2/edge channel on the same model

> We will use the same configuration as the existing synapse charm.

Reviewer: Mention that deploying the new synapse charm with the name `server-mas`

Author: updated!

```
juju deploy synapse server-mas --channel=2/edge \
    --config server_name=<server_name> \
    --config public_baseurl=<public_baseurl>
```

> ### Integrate with postgresql

Reviewer: ```suggestion
### Integrate with `postgresql` charm
```

> Integrate with the existing postgresql charm to set up the database for the new charm.

Reviewer: ```suggestion
Integrate with the existing `postgresql` charm to set up the database for the new charm.
```
Also might be worth mentioning that you need both endpoints

```
juju integrate server-mas:mas-database postgresql-k8s
juju integrate server-mas:database postgresql-k8s
```

> ### Configure OIDC

Reviewer 2: Some note about external access needed?

Author: We added the proxy config to the MAS's pebble layer so I don't think we need to do anything manually anymore

> This step requires the configuration of a client on the OIDC provider of your choice. Please look at the ["Configure an upstream SSO provider"](https://element-hq.github.io/matrix-authentication-service/setup/sso.html) section of the MAS documentation for detailed explanation and sample configurations for popular OIDC providers.

Reviewer: Add another sentence starting with `For example` that indicates this is not the only way and this is for integrating with Google

Author: updated!

```
juju deploy oauth-external-idp-integrator oidc --channel=latest/edge
juju config oidc \
  issuer_url=<issuer_url> \
  authorization_endpoint=<authorization_endpoint> \
  userinfo_endpoint=<userinfo_endpoint> \
  token_endpoint=<token_endpoint> \
  introspection_endpoint=<introspection_endpoint> \
  jwks_endpoint=<jwks_endpoint> \
  client_id=<client_id> \
  client_secret=<client_secret>

> juju relate server-mas oidc

Reviewer: ```suggestion
juju integrate server-mas oidc
```

```

To create the client, it's important to first retrieve the MAS redirect URL. First we get the upstream OIDC provider ID (`<provider_id>`) generated by the charm:
```
juju ssh --container synapse server-mas "cat /mas/config.yaml | grep 'id:'"
```

> The MAS redirect URL will be of this format: `<public_baseurl>/auth/upstream/callback/<provider_id>`. Once the client has been created, take note of the `client_id` and `client_secret` generated by the provider. The required endpoints as shown below in the charm configuration can usually be found at `<issuer_url>/.well-known/openid-configuration` 

Reviewer: ```suggestion
The MAS redirect URL will be formatted like `<public_baseurl>/auth/upstream/callback/<provider_id>`. Once the client has been created, take note of the `client_id` and `client_secret` generated by the provider. The required endpoints as shown below in the charm configuration can usually be found at `<issuer_url>/.well-known/openid-configuration`. 
```

> ## Backup the existing synapse charm and restore on the new synapse charm

Reviewer: ```suggestion
## Back up the existing synapse charm and restore on the new synapse charm
```

> ### Backup and restore using the s3-integrator charm

Reviewer: ```suggestion
### Back up and restore using the s3-integrator charm
```

> To migrate everything persisted in the Synapse's `/data` directory, we will use the `s3-integrator` charm. For more information please look at the ["Backup and Restore"](https://discourse.charmhub.io/t/synapse-docs-backup-and-restore/13212) section of the charm's documentation. First we deploy the `s3-integrator` charm and configure it with the credentials for the target s3 bucket:

Reviewer: ```suggestion
To migrate everything persisted in the Synapse's `/data` directory, we will use the `s3-integrator` charm. For more information please look at the ["Back up and Restore"](https://charmhub.io/synapse/docs/how-to-backup-and-restore) section of the charm's documentation. First we deploy the `s3-integrator` charm and configure it with the credentials for the target s3 bucket:
```

```
juju deploy s3-integrator --channel edge
juju config s3-integrator endpoint=<s3 endpoint> bucket=<bucket name> path=<optional-path> region=<region> s3-uri-style=<path or host>
juju run s3-integrator/leader sync-s3-credentials access-key=<access-key> secret-key=<secret-key>
```

> Then we integrate with the existing synapse charm and create the backup

Reviewer: ```suggestion
Then we integrate with the existing synapse charm and create the backup:
```

```

> juju relate s3-integrator server

Reviewer: ```suggestion
juju integrate s3-integrator server
```

juju run server/leader create-backup
```

> Take note of the backup ID, we will use it to restore the backup on the new synapse charm.

Reviewer: ```suggestion
Take note of the backup ID; we will use it to restore the backup on the new synapse charm.
```

```

> juju relate s3-integrator server-mas

Reviewer: ```suggestion
juju integrate s3-integrator server-mas
```

juju run synapse/leader restore-backup backup-id=<backup-id>
```

## Synapse database migration

> ### Stop all services on the new and the existing synapse charm

Reviewer: ```suggestion
### Stop all services on the new and the existing synapse charms
```

```
juju ssh --container synapse server-mas/0 "pebble stop synapse; pebble stop synapse-mas; pebble stop stats-exporter"
juju ssh --container synapse server/0 "pebble stop synapse; pebble stop stats-exporter"
```

### Connect to the database
```
juju ssh --container synapse server-mas/0 bash
```

> The following commands assume that you are in the `synapse` container of the `server-mas` application.

Reviewer: Nitpick: Might be good as a note/warning admonition

```
apt update
apt install wget postgresql-client -y
wget https://github.com/mikefarah/yq/releases/download/v4.44.3/yq_linux_amd64 -O /usr/bin/yq && chmod +x /usr/bin/yq
```

> Copy the original synapse database

Reviewer: ```suggestion
Copy the original synapse charm database:
```

Reviewer: @Thanhphan1147 besides this missing "charm"/colon, I have no further comments!

```
DB_HOST=$(yq e '.database.args.host' /data/homeserver.yaml)
DB_PORT=$(yq e '.database.args.port' /data/homeserver.yaml)
DB_PASSWORD=$(yq e '.database.args.password' /data/homeserver.yaml)
DB_USER=$(yq e '.database.args.user' /data/homeserver.yaml)
PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres -c 'drop database "server-mas";'

> PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres -c "create database \"server-mas\" with template server owner $DB_USER;"

Reviewer 2: Was the idea here to restore the current database to the new one? This seems that is just creating.

Author: The `with template server` will actually make a copy of the current database iiuc.

```

> In case where postgres gives you an error saying "", drop all connections from the database

Reviewer: ```suggestion
In case where PostgreSQL gives you an error saying "", drop all connections from the database:
```
Also update the error message

```
PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres -c SELECT pg_terminate_backend(pg_stat_activity.pid) FROM pg_stat_activity WHERE pg_stat_activity.datname = "server" AND pid <> pg_backend_pid();
```

> ## Migrating users to MAS

Reviewer: ```suggestion
## Migrate users to MAS
```

> ### Perform user migration to MAS database

Reviewer 2: a note about proxy?

Author: I thought about it but I don't think that it should be put here since it's quite specific to the environment, what do you think?

Reviewer 2: Ok, if needed can be added later

Run the `advisir` command as well as the `migrate` command with the `--dryRun` flag to check that there are no errors.
```
npx --yes @vector-im/syn2mas \
    --command advisor \
    --synapseConfigFile /data/homeserver.yaml

OAUTH2_PROVIDER_ID=$(yq e '.upstream_oauth2.providers[0].id' /mas/config.yaml)
npx --yes @vector-im/syn2mas \
    --command migrate \
    --synapseConfigFile /data/homeserver.yaml \
    --masConfigFile /mas/config.yaml \
    --upstreamProviderMapping saml:$OAUTH2_PROVIDER_ID \
    --dryRun
```

> 

Reviewer: Add some indication of what a "no error" message looks like for all the commands

Author: added, thanks!

> Finally, run the migration command

Reviewer: ```suggestion
Finally, run the migration command:
```

```
npx --yes @vector-im/syn2mas \
    --command migrate \
    --synapseConfigFile /data/homeserver.yaml \
    --masConfigFile /mas/config.yaml \
    --upstreamProviderMapping saml:$OAUTH2_PROVIDER_ID
```

> 

Reviewer: Also good to add an indication of what success looks like in the terminal

Author: added, thanks!

### Configure the OIDC subject claim

> After the migration, the `external_id` of the synapse user is used as the OIDC subject in MAS. Most of the time this needs to be changed so that the local account gets recognized by MAS when authenticating with the upstream provider. 

Reviewer: ```suggestion
After the migration, the `external_id` of the Synapse user is used as the OIDC subject in MAS. Most of the time, this needs to be changed so that the local account gets recognized by MAS when authenticating with the upstream provider. 
```
Add another sentence explaining what the `psql` command is doing (logging into the database)

```
psql $(yq e '.database.uri' /mas/config.yaml)
```

Then run the appropriate SQL command to update the subject field. For example, to use the email address as the OIDC subject, we run this SQL query:
```
update upstream_oauth_links
SET subject=(select email from user_emails where upstream_oauth_links.user_id=user_emails.user_id);

\q
```

> After updating the OIDC subject in the MAS database, we need to also configure the OIDC subject claim:

Reviewer: ```suggestion
After updating the OIDC subject in the MAS database, we need to also configure the OIDC subject claim outside of the PostgreSQL container:
```

Author: updated, thanks

# How to migrate Synapse to the `2/edge` track

> This document shows how to migrate a legacy synapse homeserver (`1/edge` track) to an oidc-native homeserver (`2/edge` track).

Reviewer: ```suggestion
This document shows how to migrate a legacy Synapse homeserver (`1/edge` track) to an oidc-native homeserver (`2/edge` track).
```

## Initial deployment setup 

> This document will cover the migration path for a synapse charm on the `1/edge` track that us using the `saml-integrator` charm to provide authentication via SAML.

Reviewer: ```suggestion
This document will cover the migration path for a synapse charm on the `1/edge` track that uses the `saml-integrator` charm to provide authentication via SAML.
```

## Prepare the new synapse charm
### Deploy the 2/edge channel on the same model
We will deploy the new synapse charm with the name `server-mas`. We will use the same configuration as the existing synapse charm.
```
juju deploy synapse server-mas --channel=2/edge \
    --config server_name=<server_name> \
    --config public_baseurl=<public_baseurl>
```

### Integrate with the `postgresql-k8s` charm

> Integrate with the existing `postgresql-k8s` charm to set up the database for the new charm. Note that you need to integrate with the `postgresql-k8s` charm via both the `database` and the `mas-database` endpoint

Reviewer: ```suggestion
Integrate with the existing `postgresql-k8s` charm to set up the database for the new charm. Note that you need to integrate with the `postgresql-k8s` charm via both the `database` and the `mas-database` endpoints.
```

```
juju integrate server-mas:mas-database postgresql-k8s
juju integrate server-mas:database postgresql-k8s
```

### Configure OIDC
This step requires the configuration of a client on the OIDC provider of your choice. Please look at the ["Configure an upstream SSO provider"](https://element-hq.github.io/matrix-authentication-service/setup/sso.html) section of the MAS documentation for detailed explanation and sample configurations for popular OIDC providers. For example, to use the Google IdP, we will deploy the `oauth-external-idp-integrator` charm.

> Once the client has been created, take note of the `client_id` and `client_secret` generated by the provider. The required endpoints as shown below in the charm configuration can usually be found at `<issuer_url>/.well-known/openid-configuration`

Reviewer: ```suggestion
Once the client has been created, take note of the `client_id` and `client_secret` generated by the provider. The required endpoints as shown below in the charm configuration can usually be found at `<issuer_url>/.well-known/openid-configuration`.
```

```
juju deploy oauth-external-idp-integrator oidc --channel=latest/edge
juju config oidc \
  issuer_url=<issuer_url> \
  authorization_endpoint=<authorization_endpoint> \
  userinfo_endpoint=<userinfo_endpoint> \
  token_endpoint=<token_endpoint> \
  introspection_endpoint=<introspection_endpoint> \
  jwks_endpoint=<jwks_endpoint> \
  client_id=<client_id> \
  client_secret=<client_secret>
juju integrate server-mas oidc
```

> Finally, we need to whitelist the MAS redirect URL on the OIDC client. First we get the upstream OIDC provider ID (`<provider_id>`) generated by the charm:

Reviewer: Is there another (more inclusive) word we can use instead of whitelist? Here's a suggestion, let me know if it still has the same meaning.
```suggestion
Finally, we need to allow the MAS redirect URL on the OIDC client. First we get the upstream OIDC provider ID (`<provider_id>`) generated by the charm:
```

```
juju ssh --container synapse server-mas "cat /mas/config.yaml | grep 'id:'"
```

> The MAS redirect URL will be formatted like `<public_baseurl>/auth/upstream/callback/<provider_id>`. We can now go back to the OIDC client configuration on the IdP and whitelist the redirect URL. 

Reviewer: Same comment here about "whitelist". Another suggestion, let me know if any meaning is lost.
```suggestion
The MAS redirect URL will be formatted like `<public_baseurl>/auth/upstream/callback/<provider_id>`. We can now go back to the OIDC client configuration on the IdP and add the redirect URL to the allowlist. 
```

## Back up the existing synapse charm and restore on the new synapse charm
### Back up and restore using the s3-integrator charm

> To migrate everything persisted in the Synapse's `/data` directory, we will use the `s3-integrator` charm. For more information please look at the ["Back up and Restore"](https://charmhub.io/synapse/docs/how-to-backup-and-restore) section of the charm's documentation. First we deploy the `s3-integrator` charm and configure it with the credentials for the target s3 bucket:

Reviewer: ```suggestion
To migrate everything persisted in the Synapse's `/data` directory, we will use the `s3-integrator` charm. For more information please look at the ["Back up and Restore"](https://charmhub.io/synapse/docs/how-to-backup-and-restore) section of the charm's documentation. First we deploy the `s3-integrator` charm and configure it with the credentials for the target S3 bucket:
```

```
juju deploy s3-integrator --channel edge
juju config s3-integrator endpoint=<s3 endpoint> bucket=<bucket name> path=<optional-path> region=<region> s3-uri-style=<path or host>
juju run s3-integrator/leader sync-s3-credentials access-key=<access-key> secret-key=<secret-key>
```

Then we integrate with the existing synapse charm and create the backup:
```
juju integrate s3-integrator server
juju run server/leader create-backup
```

Take note of the backup ID; we will use it to restore the backup on the new synapse charm.
```
juju integrate s3-integrator server-mas
juju run synapse/leader restore-backup backup-id=<backup-id>
```

## Synapse database migration

> ### Stop all services on the new and the existing synapse charms

Reviewer: Nitpick: I think this could be demoted to text like in the following suggestion:
```suggestion
Stop all services on the new and the existing synapse charms:
```

```
juju ssh --container synapse server-mas/0 "pebble stop synapse; pebble stop synapse-mas; pebble stop stats-exporter"
juju ssh --container synapse server/0 "pebble stop synapse; pebble stop stats-exporter"
```

> ### Connect to the database

Reviewer: Nitpick: I also think this header could be demoted to text.
```suggestion
Connect the new synapse charm to the database:
```

```
juju ssh --container synapse server-mas/0 bash
```

The following commands assume that you are in the `synapse` container of the `server-mas` application.
```
apt update
apt install wget postgresql-client -y
wget https://github.com/mikefarah/yq/releases/download/v4.44.3/yq_linux_amd64 -O /usr/bin/yq && chmod +x /usr/bin/yq
```

Copy the original synapse database
```
DB_HOST=$(yq e '.database.args.host' /data/homeserver.yaml)
DB_PORT=$(yq e '.database.args.port' /data/homeserver.yaml)
DB_PASSWORD=$(yq e '.database.args.password' /data/homeserver.yaml)
DB_USER=$(yq e '.database.args.user' /data/homeserver.yaml)
PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres -c 'drop database "server-mas";'
PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres -c "create database \"server-mas\" with template server owner $DB_USER;"
```

In case where postgres gives you an error saying "ERROR:  source database "synapse" is being accessed by other users", drop all connections from the database:

> First login to the database

Reviewer: ```suggestion
First log in to the database:
```

```
PGPASSWORD=$DB_PASSWORD psql --host $DB_HOST --username $DB_USER --port $DB_PORT postgres
```

> Then run this SQL query to drop all connection from the relevant database ( `<database_name>` )

Reviewer: ```suggestion
Then run this SQL query to drop all connection from the relevant database ( `<database_name>` ):
```

```
SELECT pg_terminate_backend(pg_stat_activity.pid) FROM pg_stat_activity 
WHERE pg_stat_activity.datname = '<database_name>' AND pid <> pg_backend_pid();
```

## Migrating users to MAS
### Perform user migration to MAS database
Run the `migrate` command with the `--dryRun` flag to check that there are no errors:
```
OAUTH2_PROVIDER_ID=$(yq e '.upstream_oauth2.providers[0].id' /mas/config.yaml)
npx --yes @vector-im/syn2mas \
    --command migrate \
    --synapseConfigFile /data/homeserver.yaml \
    --masConfigFile /mas/config.yaml \
    --upstreamProviderMapping saml:$OAUTH2_PROVIDER_ID \
    --dryRun
```

You should see the 0 fatals and 0 warnings in the command output:
```
[INFO] migrate - Completed migration dry-run of x users with 0 fatals and 0 warnings:
```

Finally, run the migration command:
```
npx --yes @vector-im/syn2mas \
    --command migrate \
    --synapseConfigFile /data/homeserver.yaml \
    --masConfigFile /mas/config.yaml \
    --upstreamProviderMapping saml:$OAUTH2_PROVIDER_ID
```

> You should see an output similar to that of the dry-run command

Reviewer: ```suggestion
You should see an output similar to that of the dry-run command.
```
